### DISCUSSION TITLE
Support custom Knex connection function

### DISCUSSION THREAD

**Discussion by u12206050** [directus/directus#17378] [Feature Requests]
Access token for postgress needs to be refreshed and fetched again from IdentityServer (could be azure, google, env0 etc).

It seems knex already supports this: https://knexjs.org/guide/#configuration-options
but would need a way to support it in Directus as well.

I suggest adding a hook for "db_connection" that should return a callback function that will be passed on to the knex `connection` property.

**Comment by u12206050** [comment]
Side note, this is not just useful for postgres but any cloud hosted database as well that support token authorization.

**Comment by adelinn** [comment]
Implementing this feature would enable token authentication for PostgreSQL and would enable the use of Azure Managed Identities to authenticate to the DB. https://learn.microsoft.com/en-us/azure/postgresql/flexible-server/how-to-connect-with-managed-identity
Google Cloud and AWS support Token based authentication as well.

It's highly recommended to use token authentication in favor of secrets to reduce the risks in case exploits are found in the app.
Recovering in such a case when secrets are used usually requires taking the app offline, fixing it and rotating all the secrets. If token authentication is used, only taking the app offline and fixing it is required.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
